datacube/metadata/_stacconverter.py

```diff
@@ -0,0 +1,318 @@
+                target=urljoin(stac_url, f"/stac/collections/{dataset.product.name}"),
+            )
+        yield Link(
+            title="ODC Product Overview",
+            rel="product_overview",
+            media_type="text/html",
+            target=urljoin(stac_url, f"product/{dataset.product.name}"),
+        )
+        yield Link(
+            title="ODC Dataset Overview",
+            rel="alternative",
+            media_type="text/html",
+            target=urljoin(stac_url, f"dataset/{dataset.id}"),
+        )
+
+    if not collection_url and not stac_url:
+        warnings.warn("No collection provided for Stac Item.")
+
+
+def _as_stac_instruments(value: str):
+    """
+    >>> _as_stac_instruments("TM")
+    ['tm']
+    >>> _as_stac_instruments("OLI")
+    ['oli']
+    >>> _as_stac_instruments("ETM+")
+    ['etm']
+    >>> _as_stac_instruments("OLI_TIRS")
+    ['oli', 'tirs']
+    """
+    return [i.strip("+-").lower() for i in value.split("_")]
+
+
+def _convert_value_to_stac_type(key: str, value):
+    """
+    Convert return type as per STAC specification
+    """
+    # In STAC spec, "instruments" have [String] type
+    if key == "eo:instrument":
+        return _as_stac_instruments(value)
+    # Convert the non-default datetimes to a string
+    elif isinstance(value, datetime.datetime) and key != "datetime":
+        return datetime_to_str(value)
+    else:
+        return value
+
+
+def eo3_to_stac_properties(dataset: Dataset, title: str | None = None) -> dict:
+    """
+    Convert EO3 properties dictionary to the Stac equivalent.
+    """
+    # Put the title at the top for document readability.
+    properties = {"title": title} if title else {}
+
+    properties.update(
+        {
+            EO3_TO_STAC_RENAMES.get(key, key): _convert_value_to_stac_type(key, val)
+            for key, val in dataset.properties.items()
+        },
+    )
+
+    return properties
+
+
+def ds2stac(
+    dataset: Dataset,
+    stac_url: str | None = None,
+    self_url: str | None = None,
+    collection_url: str | None = None,
+) -> Item:
+    if dataset.extent is not None:
+        wgs84_geometry = dataset.extent.to_crs("EPSG:4326", math.inf)
+
+        geometry = wgs84_geometry.json
+        bbox = wgs84_geometry.boundingbox.bbox
+    else:
+        geometry = None
+        bbox = None
+
+    properties = eo3_to_stac_properties(dataset, title=dataset.metadata.label)
+    properties.update(_lineage_fields(dataset))
+
+    dt = properties.get("datetime")
+
+    item = Item(
+        id=str(dataset.id),
+        datetime=dt,
+        properties=properties,
+        geometry=geometry,
+        bbox=bbox,
+        collection=dataset.product.name,
+    )
+
+    # Add links
+    for link in _stac_links(dataset, stac_url, self_url, collection_url):
+        item.links.append(link)
+
+    EOExtension.ext(item, add_if_missing=True)
+
+    if dataset.extent:
+        proj = ProjectionExtension.ext(item, add_if_missing=True)
```

Comment:
Will something else fail if the check in the next line fails and this line hasn't run, or can this run after the "input validation"?

Reply:
The first few lines of the function have an if-statement about `dataset.extent`, and then there's 20 lines of code doing things, and then there's a check on `dataset.crs` followed by raising an exception. Wouldn't it be faster/easier to raise the exception in the if-statement at the beginning of the function before running the 20 lines of code doing things?